File: sdk/go/ledger.go
package meridian

import "context"

// LedgerClient wraps services/ledger (api/ledger.yaml). Amounts are kobo.
type LedgerClient struct{ c *Client }

// Ledger returns a ledger client over the shared core.
func (c *Client) Ledger() *LedgerClient { return &LedgerClient{c} }

type Account struct {
	ID       string `json:"id"`
	Ledger   uint64 `json:"ledger"`
	Code     uint16 `json:"code"`
	Flags    uint16 `json:"flags,omitempty"`
	UserData string `json:"user_data,omitempty"`
}

type Balance struct {
	AccountID      string `json:"account_id"`
	DebitsPending  uint64 `json:"debits_pending"`
	DebitsPosted   uint64 `json:"debits_posted"`
	CreditsPending uint64 `json:"credits_pending"`
	CreditsPosted  uint64 `json:"credits_posted"`
	PostedNet      int64  `json:"posted_net"`
	AvailableKobo  int64  `json:"available"`
}

type Transfer struct {
	ID              string `json:"id,omitempty"`
	DebitAccountID  string `json:"debit_account_id"`
	CreditAccountID string `json:"credit_account_id"`
	Amount          uint64 `json:"amount"`
	Ledger          uint64 `json:"ledger"`
	Code            uint16 `json:"code"`
	Pending         bool   `json:"pending,omitempty"`
	Resolved        bool   `json:"resolved,omitempty"`
	UserData        string `json:"user_data,omitempty"`
}

// CreateAccounts creates accounts (idempotency key recommended).
func (l *LedgerClient) CreateAccounts(ctx context.Context, accounts []Account, idempotencyKey string) error {
	return l.c.post(ctx, "/v1/accounts", map[string]any{"accounts": accounts}, nil, idempotencyKey)
}

// Balance returns the derived balance for an account.
func (l *LedgerClient) Balance(ctx context.Context, accountID string) (*Balance, error) {
	var b Balance
	if err := l.c.get(ctx, "/v1/accounts/"+accountID+"/balance", &b); err != nil {
		return nil, err
	}
	return &b, nil
}

// Transfer posts a settled double-entry transfer.
func (l *LedgerClient) Transfer(ctx context.Context, t Transfer, idempotencyKey string) (*Transfer, error) {
	var out Transfer
	if err := l.c.post(ctx, "/v1/transfers", t, &out, idempotencyKey); err != nil {
		return nil, err
	}
	return &out, nil
}

// TransferPending creates a pending (two-phase) transfer.
func (l *LedgerClient) TransferPending(ctx context.Context, t Transfer, idempotencyKey string) (*Transfer, error) {
	var out Transfer
	if err := l.c.post(ctx, "/v1/transfers/pending", t, &out, idempotencyKey); err != nil {
		return nil, err
	}
	return &out, nil
}

// PostPending settles a pending transfer.
func (l *LedgerClient) PostPending(ctx context.Context, id string) (*Transfer, error) {
	var out Transfer
	if err := l.c.post(ctx, "/v1/transfers/"+id+"/post", nil, &out, ""); err != nil {
		return nil, err
	}
	return &out, nil
}

// VoidPending releases a pending transfer reservation.
func (l *LedgerClient) VoidPending(ctx context.Context, id string) (*Transfer, error) {
	var out Transfer
	if err := l.c.post(ctx, "/v1/transfers/"+id+"/void", nil, &out, ""); err != nil {
		return nil, err
	}
	return &out, nil
}

File: sdk/python/meridian_sdk/ledger.py
"""Ledger service client (api/ledger.yaml). Amounts are kobo."""

from __future__ import annotations

from typing import Optional

from pydantic import BaseModel

from .client import Client


class Account(BaseModel):
    id: str
    ledger: int
    code: int
    flags: int = 0
    user_data: str = ""


class Balance(BaseModel):
    account_id: str
    debits_pending: int = 0
    debits_posted: int = 0
    credits_pending: int = 0
    credits_posted: int = 0
    posted_net: int = 0
    available: int = 0


class Transfer(BaseModel):
    id: str = ""
    debit_account_id: str
    credit_account_id: str
    amount: int
    ledger: int
    code: int
    pending: bool = False
    resolved: bool = False
    user_data: str = ""


class LedgerClient:
    def __init__(self, c: Client):
        self._c = c

    def create_accounts(self, accounts: list[Account], *, idempotency_key: Optional[str] = None) -> None:
        self._c.post("/v1/accounts",
                     {"accounts": [a.model_dump(mode="json") for a in accounts]},
                     idempotency_key=idempotency_key)

    def balance(self, account_id: str) -> Balance:
        return self._c.get(f"/v1/accounts/{account_id}/balance", model=Balance)

    def transfer(self, t: Transfer, *, idempotency_key: Optional[str] = None) -> Transfer:
        return self._c.post("/v1/transfers", t, model=Transfer, idempotency_key=idempotency_key)

    def transfer_pending(self, t: Transfer, *, idempotency_key: Optional[str] = None) -> Transfer:
        return self._c.post("/v1/transfers/pending", t, model=Transfer, idempotency_key=idempotency_key)

    def post_pending(self, transfer_id: str) -> Transfer:
        return self._c.post(f"/v1/transfers/{transfer_id}/post", model=Transfer)

    def void_pending(self, transfer_id: str) -> Transfer:
        return self._c.post(f"/v1/transfers/{transfer_id}/void", model=Transfer)
